### ISSUE TITLE
contribution-manager: No visual feedback of update progress

### ISSUE THREAD

**Issue by mark-orion** [processing/processing#4105]
The contribution-manager should give visible feedback of the update progress. Currently updates appear in the "update" tab, but when clicking "update all" nothing seems to happen although in fact the download and installation runs hidden in the background.

**Comment by iamutkarshtiwari** [comment]
@alignedleft How to reproduce the situation when a update  is available in Contribution Manager provided my installed contributions are already up-to-date ? I need to do this to add the above "visual update progress" features.

**Comment by alignedleft** [comment]
@iamutkarshtiwari Good question. Can you save an older copy of a contribution (like a library) to keep around just for testing purposes?

**Comment by iamutkarshtiwari** [comment]
@alignedleft I dont have any older copy of any contribution because I recently setup my Processing build. Where can I find an older copy?

**Comment by alignedleft** [comment]
@iamutkarshtiwari I don't think we archive contributions anywhere. You would need to check the websites of various contributors to see if they archive older versions. For example, here are older copies of the official video library:

https://github.com/processing/processing-video/releases

**Comment by joelmoniz** [comment]
@iamutkarshtiwari

Another quick way to simulate this is to install a contribution, and simply
change the integer value of the "version" field to a smaller integer in the
.properties file of the installed contribution.

**Comment by iamutkarshtiwari** [comment]
@alignedleft @joelmoniz  thanks ! :)

**Comment by Akarshit** [comment]
@alignedleft Can we make some space to show a progress bar for all the contributions when the `Update All` button is clicked as some contributions are huge and take a lot of time to download. So i think a progress bar would be better than a spinner.  
Currently the progress bar of update is shown when the contribution is selected in its respective tab after `Update All` button is clicked. Perhaps we can display it in the left most corner of the status panel.

**Comment by alignedleft** [comment]
@Akarshit This is a great idea. I can't trigger any updates on my end. If it's convenient, could you post a screenshot capturing the scenario you describe above?  Then I can propose a more specific UI treatment. Thanks.

**Comment by iamutkarshtiwari** [comment]
@alignedleft @Akarshit If "download bar" sounds more interesting, I can implement that as well.

**Comment by mark-orion** [comment]
I want to throw my 5p in for the progress bar. As it has already been said: Downloads can take time - specially on a slow connection and I appreciate the work that is going on here!

**Comment by Akarshit** [comment]
@alignedleft I was working on this for a while and came up with a couple of designs

1. Simply display the progressbar in the bottom left of the window.
![updates](https://cloud.githubusercontent.com/assets/7762605/11407640/fc42ca38-93d9-11e5-8dda-a8cc8f51b73c.png)

2. Display a circular progressbar in the status column of every contribution that is being installed/updated/removed. The circle will keep filling as the progress increases. The full progress bar is still in the status panel for the selected contribution.
![screenshot from 2015-11-26 00 45 18](https://cloud.githubusercontent.com/assets/7762605/11407655/1165429c-93da-11e5-900b-65e6cd205266.png)

There are advantages to the second method like
1. The user would know what contribution are currently being installed/updated/removed.
2. The user would know the progress of every contribution.

We can color code the circular progressbar for update/install/remove.

The above screenshot is from a initial imlementation, I will polish it once the UI is approved.

**Comment by alignedleft** [comment]
@Akarshit Thanks! I like option 2, which, if I understand it, shows the progress of each individual update as well as the overall progress (such as when "update all" is triggered, and multiple updates are being processed at once).

What would option 2 look like on the "Updates" tab? (That is, if your first mockup was adapted for option 2.)

**Comment by Akarshit** [comment]
@alignedleft Yes you are correct, it shows the progress of each individual update as well as the overall progress for the updates tab.  
For the rest of the tab it shows the progress of individual install/remove/update and when a contribution is selected the status panel looks as in option 2.

**Comment by alignedleft** [comment]
@Akarshit I think no need to see the percentage.

I think we need @jamesjgrady to chime in here, as he is defining the voice of the new UI.

**Comment by jamesjgrady** [comment]
@alignedleft @Akarshit I also like option 2. I can give you a refined comp if you still need it. Seems like the orange is not the right color and the type in Downloading is not right.

**Comment by alignedleft** [comment]
@jamesjgrady A comp would be much appreciated. Thank you!

**Comment by Akarshit** [comment]
@jamesjgrady Ya sure, a comp would be great :)

**Comment by jamesjgrady** [comment]
@alignedleft @Akarshit Here is what I was thinking for the progress bars (See comp). I like the idea of little bars next to the item that is getting updated. It provides more visual contrast than the circles but is still pretty subtle. It will also relate to the progress bar for "Installing", "Updating", and "Removing". The words should change with the button that has been activated from "Install" to "Installing", "Update" to "Updating" etc... Let me know if you have any questions. Thanks, James

![151207-pf-contmgr-001](https://cloud.githubusercontent.com/assets/2600893/11636596/e64d7dda-9cea-11e5-8776-d1faad66f9b8.png)

**Comment by alignedleft** [comment]
@jamesjgrady Oooh, this is really nice. I like the redundancy. 

I understand from @benfry that we may not have full latitude in customizing the button elements (as you've done here), but the progress bars in the Status column are theoretically possible.

**Comment by Akarshit** [comment]
@jamesjgrady @alignedleft I think we might need to expand the status column a little to make the progress bar look good in that, because in linux the progressbar comes with rounded edged by default so that requires a little bit more of space to display the progress along with the percentage.
I will do a mockup soon and post it here.

**Comment by Akarshit** [comment]
@alignedleft This is the how the CM looks with the default progress bars that are provided by java. Do you want me to customize them to look similar to those in the mockup above ?
![screenshot from 2015-12-11 11 03 04](https://cloud.githubusercontent.com/assets/7762605/11736747/03546c44-9ff8-11e5-9315-4447ff94244e.png)

**Comment by alignedleft** [comment]
@Akarshit Yes, please! If this is an insane amount of work, perhaps we can live without it. But ideally, the progress bars would:

- be gray, not orange
- mirror the typography (font and sizing) used in @jamesjgrady’s mockup

Also the Status column bar would have hard corners (not rounded).

The "Install" button progress bar would ideally be implemented *within* the button itself (or temporarily replacing the button), as in the mockup. If that's not possible, at least the styling should match the Status column progress bar, as described above.

Thank you for exploring this!

**Comment by benfry** [comment]
Simple version from https://github.com/processing/processing/pull/4154 now merged for 3.0.2. Not quite the full download indicator thing, but it's a start.

**Comment by Akarshit** [comment]
@benfry Should i continue working on #4229 or you have something else in mind ?
